### ISSUE TITLE
containerize: Ubuntu fails with `update: false`

### ISSUE THREAD

**Issue by blue42u** [LLNL/spack#37772]
### Steps to reproduce

```console
$ cat spack.yaml
spack:
  specs: [zlib]
  container:
    os_packages:
      update: false
      final: [gcc]
$ spack containerize > Dockerfile
$ buildah build --format=docker
```

### Error message

The attempt to install the requested packages (`gcc`) fails because `apt-get update` was not run before `apt-get install`:
```console
$ buildah build --format=docker
...
[2/2] STEP 7/9: RUN apt-get -yqq install gcc  && rm -rf /var/lib/apt/lists/*
E: Unable to locate package gcc
error building at STEP "RUN apt-get -yqq install gcc  && rm -rf /var/lib/apt/lists/*": error while running runtime: exit status 100
```

### Information on your system

$ spack debug report
* **Spack:** 0.21.0.dev0 (c2ce9a6d930891c708d822be8ce22aa5f12a58ee)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
I'd like to think a bit more on how to solve this, or even if this should be considered a bug at all. :thinking: 

The reason is that I'm not really keen on adding a new "preinstall" command for system package managers, when we can already work around that using [this feature](https://spack.readthedocs.io/en/latest/containers.html#extending-the-jinja2-templates). The option, as you correctly noted, is mainly for `apt` and in the beginning this was conceived for custom images where people already did an update.

**Comment by blue42u** [comment]
I'm positive it's at least a bug. `apt-get update` and `apt-get install` **must** always be in the same `RUN` statement. This prevents non-intuitive issues caused by the Docker layer cache, see [the Dockerfile best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#apt-get). In that sense, the initial intention of `update: false` (meaning "skip `apt-get update`") is horribly misguided.

Right now, `update: true` means "update all installed packages" (i.e. `apt-get upgrade`). This is not what the docs say but this is the behavior in practice. By extension, `update: false` means "skip updating all installed packages" for every `os` except for `ubuntu:*`. This inconsistency makes this error seem like a bug.

As it stands right now, there is no way (except templates) to install packages in an `os: ubuntu:*` image without updating *all* installed packages (i.e. `apt-get upgrade`). I don't mind using templates, but if that's necessary I'd like to be told in the docs, not by mysterious failures very late in the container build process.

**Comment by alalazo** [comment]
> I don't mind using templates, but if that's necessary I'd like to be told in the docs, not by mysterious failures very late in the container build process.

Oh, that is for sure. What I am currently thinking is to deal with it automatically, without the need to introduce an additional configuration attribute, that seems rather arbitrary[^1]. In short, special case the template mechanism so that for `apt` with `update: false` the install command will be:
```console
apt-get update && apt-get install ...
```

[^1]: if there's `preinstall`, why not having `postinstall` etc.

**Comment by blue42u** [comment]
Agreed. That was also my initial thought, but I didn't see a clear way to special-case it without duplicating and hardcoding the names from [`images.json`](https://github.com/spack/spack/blob/develop/lib/spack/spack/container/images.json), hence the PR for `preinstall`.

Special-casing does sound attractive though, there are a number of other cases that would benefit, e.g.:
  - `dnf_epel` could be replaced with a special-case of `dnf` on `rockylinux:*` or `almalinux:*`,
  - `yum_amazon` could be replaced with a special-case of `yum` on `amazonlinux:*`,
  - `dnf` and `yum` could be aliases for each other and special-cased by the os, e.g. `yum` on `fedora:*` really means `dnf` and `dnf` on `centos:7` really means `yum`.
